### ISSUE TITLE
Invalid size of SSD

### ISSUE THREAD

**Issue by chmurli** [nokyan/resources#21]
![image](https://github.com/nokyan/resources/assets/434440/66829ae4-b4fc-462b-bae8-37f542495dcd)

SSD size is reported as 16TB (would love to have) but in fact it's 2TB.

System: Arch
SSH: Kingston KC3000 2TB

**Comment by nokyan** [comment]
Hi, thanks for pointing out an issue.
That's really odd, my only guess is that maybe Resources gets its sector size wrong? What does `cat /sys/block/nvme0n1/size` output for you?

**Comment by chmurli** [comment]
cat /sys/block/nvme0n1/size 
4000797360

FYI: other tools reports correct size.

**Comment by jthat** [comment]
@nokyan For the `size` file, the sector size seems to be hardcoded to 512 regardless of the hardware sector size.

I couldn't find this documented anywhere except the relevant code:
- In the kernel, the [`size` response](https://github.com/torvalds/linux/blob/dd72f9c7e512da377074d47d990564959b772643/block/genhd.c#L1045) calls [`part_size_show()`](https://github.com/torvalds/linux/blob/dd72f9c7e512da377074d47d990564959b772643/block/genhd.c#L937-L941) which calls [`bdev_nr_sectors()`](https://github.com/torvalds/linux/blob/dd72f9c7e512da377074d47d990564959b772643/include/linux/blkdev.h#L765-L768) which returns `sector_t` which [is in units of 512](https://github.com/torvalds/linux/blob/dd72f9c7e512da377074d47d990564959b772643/include/linux/blk_types.h#L23-L28).
- In `lsblk`, this value is also hardcoded when [the `size` file is read to convert it to bytes](https://github.com/util-linux/util-linux/blob/f3fd64111c28de1ecd47f7cd519120ffa531800d/misc-utils/lsblk.c#L1371-L1372).

Therefore the [sector size calculation](https://github.com/nokyan/resources/blob/632e6717b02afc46f9961133fbd4305599fab53a/src/ui/pages/drive.rs#L267) using [hw_sector_size](https://github.com/nokyan/resources/blob/632e6717b02afc46f9961133fbd4305599fab53a/src/ui/pages/drive.rs#L210) initialized via [sector_size()](https://github.com/nokyan/resources/blob/632e6717b02afc46f9961133fbd4305599fab53a/src/utils/drive.rs#L235-L241) appears to be incorrect. If reading `queue/hw_sector_size` reports 4096, then this would explain the 8x discrepancy.

**Comment by nokyan** [comment]
> @nokyan For the `size` file, the sector size seems to be hardcoded to 512 regardless of the hardware sector size.
> 
> I couldn't find this documented anywhere except the relevant code:
> - In the kernel, the [`size` response](https://github.com/torvalds/linux/blob/dd72f9c7e512da377074d47d990564959b772643/block/genhd.c#L1045) calls [`part_size_show()`](https://github.com/torvalds/linux/blob/dd72f9c7e512da377074d47d990564959b772643/block/genhd.c#L937-L941) which calls [`bdev_nr_sectors()`](https://github.com/torvalds/linux/blob/dd72f9c7e512da377074d47d990564959b772643/include/linux/blkdev.h#L765-L768) which returns `sector_t` which [is in units of 512](https://github.com/torvalds/linux/blob/dd72f9c7e512da377074d47d990564959b772643/include/linux/blk_types.h#L23-L28).
> - In `lsblk`, this value is also hardcoded when [the `size` file is read to convert it to bytes](https://github.com/util-linux/util-linux/blob/f3fd64111c28de1ecd47f7cd519120ffa531800d/misc-utils/lsblk.c#L1371-L1372).
> 
> Therefore the [sector size calculation](https://github.com/nokyan/resources/blob/632e6717b02afc46f9961133fbd4305599fab53a/src/ui/pages/drive.rs#L267) using [hw_sector_size](https://github.com/nokyan/resources/blob/632e6717b02afc46f9961133fbd4305599fab53a/src/ui/pages/drive.rs#L210) initialized via [sector_size()](https://github.com/nokyan/resources/blob/632e6717b02afc46f9961133fbd4305599fab53a/src/utils/drive.rs#L235-L241) appears to be incorrect. If reading `queue/hw_sector_size` reports 4096, then this would explain the 8x discrepancy.

That is very insightful, thank you so much! I'll fix this then by also hard coding the sector size to 512

**Comment by diegorodriguezv** [comment]
Similar problem here. Resources report my external USB drive as 24TB (21.83 TiB).
```
❯ cat /sys/block/sdb/size
5860466688
❯ df
Filesystem                 1K-blocks       Used Available Use% Mounted on
...
/dev/sdb1                 2930232316 2727399900 202832416  94% /mnt/mybook
❯ lsusb
...
Bus 004 Device 002: ID 1058:1140 Western Digital Technologies, Inc. My Book Essential (WDBACW)
```
It is a Western Digital - My Book 1140 - 3TB external USB Hard Drive.
Let me know if I can be of any help. 
Thanks.

**Comment by nokyan** [comment]
Thanks for reporting, this will be fixed with Resources 1.2 which I plan to release end of next week or so (not a guarantee though) :)

**Comment by jthat** [comment]
This is fixed for me in 1.2.0. Thanks!

**Comment by nokyan** [comment]
> This is fixed for me in 1.2.0. Thanks!

Great to see!
Closing this then. If there's still something wrong, don't hesitate to reopen. :)

**Comment by diegorodriguezv** [comment]
I have verified the fix. Resources now show my external USB drive size correctly. Unfortunately it also shows about 20 loop devices (used for snaps in ubuntu). I'll open a new issue.
